File: Snooper/src/test/java/com/prateekj/snooper/networksnooper/presenter/HttpCallListPresenterTest.java
package com.prateekj.snooper.networksnooper.presenter;

import androidx.annotation.NonNull;

import com.prateekj.snooper.networksnooper.database.SnooperRepo;
import com.prateekj.snooper.networksnooper.model.HttpCall;
import com.prateekj.snooper.networksnooper.model.HttpCallRecord;
import com.prateekj.snooper.networksnooper.views.HttpListView;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class HttpCallListPresenterTest {

  @Mock
  private HttpListView view;
  @Mock
  private SnooperRepo repo;
  private HttpCallListPresenter httpCallListPresenter;

  @Before
  public void setUp() throws Exception {
    this.httpCallListPresenter = new HttpCallListPresenter(view, repo);
  }

  @Test
  public void shouldInitializeHttpCallRecordsList() throws Exception {
    List<HttpCallRecord> httpCallRecords = singletonList(mock(HttpCallRecord.class));
    when(repo.findAllSortByDateAfter(-1, 20)).thenReturn(httpCallRecords);
    httpCallListPresenter.init();

    verify(view).initHttpCallRecordList(httpCallRecords);
  }

  @Test
  public void shouldShowNoCallsFoundMessage() throws Exception {
    when(repo.findAllSortByDateAfter(-1, 20)).thenReturn(new ArrayList<HttpCallRecord>());
    httpCallListPresenter.init();

    verify(view).renderNoCallsFoundView();
    verify(view, never()).initHttpCallRecordList(any(List.class));
  }

  @Test
  public void shouldQueryNextSetOfRecordOnNextPageCall() throws Exception {
    List<HttpCallRecord> lastSetHttpCalls = asList(createCallWithId(1));
    List<HttpCallRecord> nextSetHttpCalls = asList(createCallWithId(3), createCallWithId(2));
    when(repo.findAllSortByDateAfter(-1, 20)).thenReturn(asList(createCallWithId(5), createCallWithId(4)));
    httpCallListPresenter.init();

    when(repo.findAllSortByDateAfter(4, 20)).thenReturn(nextSetHttpCalls);
    httpCallListPresenter.onNextPageCall();
    verify(repo).findAllSortByDateAfter(4, 20);
    verify(view).appendRecordList(nextSetHttpCalls);

    when(repo.findAllSortByDateAfter(2, 20)).thenReturn(lastSetHttpCalls);
    httpCallListPresenter.onNextPageCall();
    verify(repo).findAllSortByDateAfter(2, 20);
    verify(view).appendRecordList(lastSetHttpCalls);
  }

  @Test
  public void shouldNotifyViewToNavigateToResponseBody() throws Exception {
    HttpCallRecord httpCall = mock(HttpCallRecord.class);
    when(httpCall.getId()).thenReturn(2L);

    httpCallListPresenter.onClick(httpCall);

    verify(view).navigateToResponseBody(2);
  }

  @Test
  public void shouldNotifyViewToFinishIt() throws Exception {
    httpCallListPresenter.onDoneClick();

    verify(view).finishView();
  }

  @Test
  public void shouldDeleteTheRecordsAndUpdateUi() throws Exception {
    httpCallListPresenter.confirmDeleteRecords();

    verify(repo).deleteAll();
    verify(view).updateListViewAfterDelete();
  }

  @Test
  public void shouldShowConfirmationDialogOnClickOfDeleteRecords() throws Exception {
    httpCallListPresenter.onDeleteRecordsClicked();

    verify(view).showDeleteConfirmationDialog();
  }

  @NonNull
  private HttpCallRecord createCallWithId(int id) {
    HttpCallRecord firstHttpCall = HttpCallRecord.from(new HttpCall.Builder().build());
    firstHttpCall.setId(id);
    return firstHttpCall;
  }
}

File: Snooper/src/test/java/com/prateekj/snooper/networksnooper/presenter/HttpCallListPresenterTest.kt
package com.prateekj.snooper.networksnooper.presenter

import com.prateekj.snooper.networksnooper.database.SnooperRepo
import com.prateekj.snooper.networksnooper.model.HttpCall
import com.prateekj.snooper.networksnooper.model.HttpCallRecord
import com.prateekj.snooper.networksnooper.views.HttpListView
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.Before
import org.junit.Test

class HttpCallListPresenterTest {

  private lateinit var view: HttpListView
  private lateinit var repo: SnooperRepo
  private lateinit var httpCallListPresenter: HttpCallListPresenter

  @Before
  @Throws(Exception::class)
  fun setUp() {
    view = mockk(relaxed = true)
    repo = mockk(relaxed = true)
    this.httpCallListPresenter = HttpCallListPresenter(view, repo)
  }

  @Test
  @Throws(Exception::class)
  fun shouldInitializeHttpCallRecordsList() {
    val httpCallRecords = mutableListOf(mockk<HttpCallRecord>(relaxed = true))
    every { repo.findAllSortByDateAfter(-1, 20) } returns httpCallRecords
    httpCallListPresenter.init()

    verify { view.initHttpCallRecordList(httpCallRecords) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldShowNoCallsFoundMessage() {
    every { repo.findAllSortByDateAfter(-1, 20) } returns mutableListOf()
    httpCallListPresenter.init()

    verify { view.renderNoCallsFoundView() }
    verify(exactly = 0) { view.initHttpCallRecordList(any()) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldQueryNextSetOfRecordOnNextPageCall() {
    val lastSetHttpCalls = mutableListOf(createCallWithId(1))
    val nextSetHttpCalls = mutableListOf(createCallWithId(3), createCallWithId(2))
    every { repo.findAllSortByDateAfter(-1, 20) } returns mutableListOf(
      createCallWithId(5),
      createCallWithId(4)
    )

    httpCallListPresenter.init()

    every { repo.findAllSortByDateAfter(4, 20) } returns nextSetHttpCalls
    httpCallListPresenter.onNextPageCall()
    verify { repo.findAllSortByDateAfter(4, 20) }
    verify { view.appendRecordList(nextSetHttpCalls) }

    every { repo.findAllSortByDateAfter(2, 20) } returns lastSetHttpCalls
    httpCallListPresenter.onNextPageCall()
    verify { repo.findAllSortByDateAfter(2, 20) }
    verify { view.appendRecordList(lastSetHttpCalls) }

    every { repo.findAllSortByDateAfter(1, 20) } returns mutableListOf()
    httpCallListPresenter.onNextPageCall()
    verify { repo.findAllSortByDateAfter(1, 20) }
    verify { view.appendRecordList(match { it.isEmpty() }) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldNotifyViewToNavigateToResponseBody() {
    val httpCall = mockk<HttpCallRecord>(relaxed = true)
    every { httpCall.id } returns 2L

    httpCallListPresenter.onClick(httpCall)

    verify { view.navigateToResponseBody(2) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldNotifyViewToFinishIt() {
    httpCallListPresenter.onDoneClick()

    verify { view.finishView() }
  }

  @Test
  @Throws(Exception::class)
  fun shouldDeleteTheRecordsAndUpdateUi() {
    httpCallListPresenter.confirmDeleteRecords()

    verify { repo.deleteAll() }
    verify { view.updateListViewAfterDelete() }
  }

  @Test
  @Throws(Exception::class)
  fun shouldShowConfirmationDialogOnClickOfDeleteRecords() {
    httpCallListPresenter.onDeleteRecordsClicked()

    verify { view.showDeleteConfirmationDialog() }
  }

  private fun createCallWithId(id: Int): HttpCallRecord {
    val firstHttpCall = HttpCallRecord.from(HttpCall.Builder().build())
    firstHttpCall.id = id.toLong()
    return firstHttpCall
  }
}